File: 680_valid_palindrome_ii.rb
=begin
Given a non-empty string s, you may delete at most one character. Judge whether you can make it a palindrome.

Example 1:
Input: "aba"
Output: True
Example 2:
Input: "abca"
Output: True
Explanation: You could delete the character 'c'.
Note:
The string will only contain lowercase characters a-z. The maximum length of the string is 50000.
=end

# @param {String} s
# @return {Boolean}
def valid_palindrome(s)
    0.upto(s.length/2-1).each do |i|
        if s[i] != s[~i]
            j = s.length - 1 - i
            return is_pali_range(s, i+1, j) || is_pali_range(s, i, j-1)
        end
    end
    
    return true
end

def is_pali_range(s, i, j)
   i.upto(j-1).each do |k|
      if s[k] != s[j-k+i]
          return false
      end
   end
    
    return true
end

File: 680_valid_palindrome_ii.rs
/*
Given a string s, return true if the s can be palindrome after deleting at most one character from it.

 

Example 1:

Input: s = "aba"
Output: true
Example 2:

Input: s = "abca"
Output: true
Explanation: You could delete the character 'c'.
Example 3:

Input: s = "abc"
Output: false
 

Constraints:

1 <= s.length <= 105
s consists of lowercase English letters.
*/

impl Solution {
    pub fn valid_palindrome(s: String) -> bool {
        Self::helper(&s.as_bytes(), true)
    }

    fn helper(chars: &[u8], can_skip: bool) -> bool {
        if chars.len() == 0 || chars.len() == 1 {
            return true;
        }

        let mut i = 0;
        let mut j = chars.len() - 1;
        while i <= j {
            if chars[i] == chars[j] {
                i += 1;
                j -= 1;
                continue;
            }
            if !can_skip {
                return false;
            }
            if Self::helper(&chars[(i + 1)..=j], false) {
                return true;
            } else if Self::helper(&chars[i..=(j - 1)], false) {
                return true;
            } else {
                return false;
            }
        }
        true
    }
}

